core/src/main/java/hivemall/embedding/Word2VecUDTF.java

```diff
@@ -0,0 +1,364 @@
+
+        if (isStringInput) {
+            this.negativeTableElementOI = HiveUtils.asStringOI(negativeTableElementListOI.getListElementObjectInspector());
+            this.wordOI = HiveUtils.asStringOI(docOI.getListElementObjectInspector());
+        } else {
+            this.negativeTableElementOI = HiveUtils.asFloatingPointOI(negativeTableElementListOI.getListElementObjectInspector());
+            this.wordOI = HiveUtils.asIntCompatibleOI(docOI.getListElementObjectInspector());
+        }
+
+        List<String> fieldNames = new ArrayList<>();
+        List<ObjectInspector> fieldOIs = new ArrayList<>();
+
+        fieldNames.add("word");
+        fieldNames.add("i");
+        fieldNames.add("wi");
+
+        if (isStringInput) {
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
+        } else {
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+        }
+
+        fieldOIs.add(PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+        fieldOIs.add(PrimitiveObjectInspectorFactory.writableFloatObjectInspector);
+
+        this.model = null;
+        this.word2index = null;
+        this.S = null;
+        this.aliasWordIds = null;
+
+        return ObjectInspectorFactory.getStandardStructObjectInspector(fieldNames, fieldOIs);
+    }
+
+    @Override
+    public void process(Object[] args) throws HiveException {
+        if (model == null) {
+            parseNegativeTable(args[0]);
+            this.model = createModel();
+        }
+
+        List<?> rawDoc = docOI.getList(args[1]);
+
+        // parse rawDoc
+        final int docLength = rawDoc.size();
+        final int[] doc = new int[docLength];
+        if (isStringInput) {
+            for (int i = 0; i < docLength; i++) {
+                doc[i] = getWordId(PrimitiveObjectInspectorUtils.getString(rawDoc.get(i), wordOI));
+            }
+        } else {
+            for (int i = 0; i < docLength; i++) {
+                doc[i] = PrimitiveObjectInspectorUtils.getInt(rawDoc.get(i), wordOI);
+            }
+        }
+
+        model.trainOnDoc(doc);
+    }
+
+    @Override
+    protected Options getOptions() {
+        Options opts = new Options();
+        opts.addOption("n", "numTrainWords", true,
+            "The number of words in the documents. It is used to update learning rate");
+        opts.addOption("dim", "dimension", true, "The number of vector dimension [default: 100]");
+        opts.addOption("win", "window", true, "Context window size [default: 5]");
+        opts.addOption("neg", "negative", true,
+            "The number of negative sampled words per word [default: 5]");
+        opts.addOption("iter", "iteration", true, "The number of iterations [default: 5]");
+        opts.addOption("model", "modelName", true,
+            "The model name of word2vec: skipgram or cbow [default: skipgram]");
+        opts.addOption(
+            "lr",
```

Comment:
consistent naming `eta0`, `learningRate` for the initial learning rate.

Reply:
I see.
Does `longOpt` remain `learningRate` or remove this field?

Reply:
remain `learningRate` for longOpt and use `eta0` for initialLearningRate.